torchao/dtypes/nf4tensor.py

```diff
@@ -160,7 +192,7 @@ def __new__(
             tensor_meta.original_shape,
             tensor_meta.original_strides,
             tensor_meta.storage_offset,
-            dtype=tensor_meta.dtype,
+            dtype=torch.bits2x4,
```

Comment:
for provenance I still dont like this 🙃

Reply:
I agree. We need a better extensibility story for dtypes.

Reply:
yeah I think we want to deprecate these, why not use torch.uint2?

Reply:
nf4 is a 4bit type. I suppose another mitigation is a type guard at torch_dispatch level and using torch.bits8 just so the allocator will always spit out bytes (not like it has a choice).

Reply:
torch.bits2x4 means 8 bit though, these dtypes (including bits1x8, bits4x2) should be removed actually, since torch.bits8 means the same thing because the meaning is uninterpreted dtypes

so what are you trying to express here? 2 bits * 2 that packed into a 4 bit?

Reply:
The concrete layout of the data for an NF4Tensor is described here: https://github.com/pytorch-labs/ao/blob/5e48e18ac4ed37f03172a8b36ddd6ac62e9ba5eb/torchao/dtypes/nf4tensor.py#L377-L379

Two parameters are packed into 1 byte. So in my mind if you wanted to use bits8 it should be on the self.quantized_data- I use a uint8 but it  the quantized data is not actually representing an unsigned int.

If we were to do this then during "dequant/get_original_weight" you need to call `self.quantized_data.view(torch.uint8)` and then do the bit shifting since that is not defined on tensors of type bits8 AFAIK.

If I wrote a kernel to do dequant+matmul I would probably use bits8 since I don't need any pytorch ops do actually be run on uint8.

However, I still think that the nf4tensor should broadcast to the world the type it is trying to emulate, and the types of the internal tensors can be whatever they need to be.

Reply:
this sounds like a uint4 tensor with a different packing format, can you reuse uint4 Tensor as the underlying dtype (by inheriting from UInt4Tensor probably)? can you write down all the use cases for nf4 dtype as well so we get some idea of how we can support it?

bits8 is generally not recommended right now either btw, since all these bit shifting ops etc. are already available in uint8 so we'd recommend uint8 if you want a 8 bit dtype.

Reply:
> for provenance I still dont like this 🙃.  I think that nf4tensor's outer wrapper subclass should have the same dtype as the type that it was created from.

I agree.  Having this represent the high precision dtype has worked well for `Float8Tensor`.

Reply:
Yeah Uint4Tensor is the same as NF4Tensor, AFAIK I think ed copied the packing format from NFTensor in nuggets and that was the basis of nf4tensor. It was copied over and not inherited for speed of enabling torchtune. 

That being said this same outer tensor dtype issue applies the same for the uint4tensor same as it does this

Reply:
So Float8Tensor's dtype is bfloat16?

Reply:
yes.  For float8 tensor specifically, this is required, because we need to trick autograd's `assert x.dtype == x.grad.dtype` restriction.  But it's also conceptually simple to reason about, "this is an emulation of a bfloat16 tensor with scaled float8".